torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,278 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
       TransformRFactor::runReplay2(this, axes)};
 }
 
+namespace {
+
+/** \brief Container class DisjointSet models equivalence relationships
+ *  \details each instance of this class keeps a set of equivalent classes
+ *           DisjointSet::join(a,b) makes the full class of a and b equivalent
+ *           DisjointSet::areEqual(a,b) checks if a and b belong same class
+ *
+ */
+template <typename T>
+class DisjointSet {
+ public:
+  DisjointSet() = default;
+
+  /** \brief Joins the equivalent class that a and b belong to
+   *         areEqual(a',b') will be true for each a'=a and b'=b
+   */
+  void join(T a, T b) {
+    int i0 = fixedPoint(a), i1 = fixedPoint(b);
+    int new_parent, new_child;
+
+    // either order here is correct but joining larger class to smaller class
+    // tend to be faster
+    std::tie(new_parent, new_child) = (weights[i0] < weights[i1])
+        ? std::make_pair(i0, i1)
+        : std::make_pair(i1, i0);
+    weights[new_parent] += weights[new_child];
+    setMap[new_child] = new_parent;
+  }
+
+  /** \brief Checks if a and b belong to the same equivalent class */
+  bool areEqual(T a, T b) {
```

Comment:
It's a little counter-intuitive that this is not const. If I understand it correctly, this is due to the small pointer-chasing optimization in `fixedPoint`. I wonder how much the optimization would give us and giving up more common const-model would be considered reasonable.

Reply:
Actually they should be constant. Thanks for pointing this out. Updated in the new commit.

I have added in the doc string that the template type T is assumed to be some hash-able value, in my use case I used Val*. So it actually makes sense to make them const.

Reply:
Actually they could be const, but as @tlemo suggested I will keep them not constant as they are parameters.

Reply:
I think what @tlemo meant is that making primitive types constant doesn't buy us anything. That's I understand. However, here, `T` is `Val*`, so `const Val*` and `Val*` have meaningful difference.

Reply:
Yes, I agree that adding const does have a difference. This container class only requires T to be hash-able.

Reply:
My original comment here actually refers to the constness of the function itself, i.e., `bool areEqual(T a, T b) const`. Why comparing some two values implies changes of DisjointSet?